### PR TITLE
The last matrix operation

### PR DESCRIPTION
**Pull request by phyboyzhang**
This pr handles the data matrix transform between GENE's datas and XGC's datas due to the field-aligned approach and Fourier transform. I rewrote the relevant codes in GENE by CPP, but actually can not fully understand the related algorithms. The execution of the merged codes report an error at the last  step MPI_Finalize. We will solve it.

### CODE REVIEW

File: src/dataprocess.h [hunk 1]
```diff
@@ -101,33 +125,8 @@ class DatasProc3D {
...
-    P3Data(LO li, LO lj,GO blockcount_,LO* mylk) : li0(li), lj0(lj), 
-          blockcount(blockcount_), mylk0(mylk) {};
-      const LO li0;
-      const LO lj0;
-      const GO blockcount;
-      LO const* const mylk0;
-  } p3;
-
+  Part1ParalPar3D p1;
+  Part3Mesh3D p3;
```
Comment (cwsmith) [inline src/dataprocess.h:129]: We should store pointers to p1 and p3 instead of by copies.

Comment (phyboyzhang) [reply src/dataprocess.h:129]: Ok,I'll change them.

**Comment by phyboyzhang** [comment]
This pr finalizes MPI successfully. 
```
0 before kokkos finalize
3 before kokkos finalize
1 before kokkos finalize
2 before kokkos finalize
2 done kokkos finalize
0 done kokkos finalize
3 done kokkos finalize
1 done kokkos finalize
MPI is finalized.
MPI is finalized.
MPI is finalized.
MPI is finalized.
```

File: src/BoundaryDescr3D.cc [hunk 2]
```diff
@@ -192,7 +192,7 @@ void BoundaryDescr3D::zPotentBoundaryBufAssign(
       }
    }
 }
-
+*/
```
Comment (cwsmith) [inline src/BoundaryDescr3D.cc:195]: Please remove these commented functions.  We can restore them from the git history if we need them.

File: src/importpart3mesh.h [hunk 3]
```diff
@@ -44,20 +44,18 @@ class Part3Mesh3D{
     LO  shiftx;  // The number of surfaces shifting from part3 first surface to part1 first surface.
 
 // paramters to create the background of mgnetic field, density and temperature profile
+/*
```
Comment (cwsmith) [inline src/importpart3mesh.h:47]: Can this commented block of variables be removed?

Comment (phyboyzhang) [reply src/importpart3mesh.h:47]: Removed.

File: src/BoundaryDescr3D.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (phyboyzhang) [reply src/BoundaryDescr3D.cc:195]: Removed.

File: src/BoundaryDescr3D.cc [hunk 4]
```diff
@@ -253,6 +245,8 @@ for(int i=0;i<lx;i++){
     }
   }
 }
+*/
```
[the remark this answers is not in the document]
Comment (phyboyzhang) [reply src/BoundaryDescr3D.cc:248]: Removed.

File: src/dataprocess.cc [hunk 5]
```diff
@@ -111,252 +104,521 @@ void DatasProc3D::AllocDensityArrays()
...
    }  
    free(tmp);
+   tmp=NULL;
 }
 
 // Assemble the density sub2d array in each process into a global one, which is straightforwardly transferred by
 // the adios2 API from coupler to Part3.
-
-void DatasProc3D::AssemDensiSendtoPart3(const Part3Mesh3D &p3m3d, const Part1ParalPar3D& p1pp3d)
+void DatasProc3D::oldAssemDensiSendtoPart3(BoundaryDescr3D& bdesc)
```
[the remark this answers is not in the document]
Comment (phyboyzhang) [reply src/dataprocess.cc:395]: The old function may be used later, as GENE keeps two versions of such kind of functions. Choosing them depends on the parameters to some extent. We may remove them or modify them when we are sure that the structure of this part of the code is  stable.

File: src/dataprocess.cc [hunk 6]
```diff
@@ -111,252 +104,521 @@ void DatasProc3D::AllocDensityArrays()
...
+    }     
+  }
+
+  for(LO i=0;i<p1->li0;i++){
+    for(LO j=0;j<p3->lj0;j++) free(loc_data[i][j]);
+    free(loc_data[i]);
   }
+  free(loc_data);
+
+//don't understand the above operation
```
[the remark this answers is not in the document]
Comment (phyboyzhang) [reply src/dataprocess.cc:521]: I actually understood it now.

File: src/dataprocess.cc [hunk 7]
```diff
@@ -111,252 +104,521 @@ void DatasProc3D::AllocDensityArrays()
...
+
+      debug=false;
+      if(debug){
+	num=0;
+	for(LO h=0;h<p1->npz;h++){
+	  num+=recvcount[h];
+	}	   
+	std::cout<<"num versurf[xl]="<<num<<" "<<p3->versurf[xl]<<'\n';
+	assert(num==p3->versurf[xl]);
+      }
```
[the remark this answers is not in the document]
Comment (phyboyzhang) [reply src/dataprocess.cc:563]: It's actually intended in the code.

File: src/dataprocess.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (cwsmith) [reply src/dataprocess.cc:521]: OK

File: src/dataprocess.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (cwsmith) [reply src/dataprocess.cc:563]: Thank you.

**Comment by cwsmith** [comment]
The manually merged code builds successfully on AiMOS.  `ctest` (`test_init`) fails.  I think having some way to test the code without running with XGC and GENE is a good idea.  An issue to discuss/fix it is here #87 .

**Pull request merged**